### DISCUSSION TITLE
questions about commands

### DISCUSSION THREAD

**Discussion by goyalyashpal** [sumatrapdfreader/sumatrapdf#3178] [Q&A] [answered]
#### Thanks for these 😊
* Thanks a lottt for the command palette, i just recently discovered it in sumatra just 3-5 days ago.
* also found out from the command palette, that sumatra also has "scroll by half screen" feature (though, im-humble-o wrongly named as half page, but anyway). i was longing for such a thing ever since i discovered it in sioyek pdf reader some weeks ago. thanks for that too...

#### Question
I wanted to ask, is there a command which:
1. while opening a new document, sets the directory to same as that of current pdf? or alternatively, some command on the lines of "Open from recent directories" where sumatra populates the recent directory list?

#### Found at other place:
1. https://github.com/sumatrapdfreader/sumatrapdf/discussions/3077#discussioncomment-4052840 by 🔍: [recent screen](https://github.com/sumatrapdfreader/sumatrapdf/discussions?discussions_q=recent+screen): 
_opening the "frequently read" home page in a new tab even when there're some documents already open?_

**Comment by GitHubRulesOK** [comment]
Hmm open files current directory was easiest as F2 (rename) then select another file to open and cancel the rename, you got both tabs HOWEVER there is a risk of unwittingly renaming current file by  premature selection so now `File Show in Folder` is safer for picking and open nearby files.

**Reply by goyalyashpal** [reply]
> so now `File Show in Folder` is safer for picking and open nearby files.

i HATE the microsoft's default file explorer (win 10), it's too inflexible and slow, and inadequate - even the QTtabbar can enhance it so much. that's why i am always looking for ways to avoid the likes of "open in explorer" options across all softwares i use.

i have tried doublecmd (double commander), mucommander, explorer++ but none of these were capable enough.

i haven't found any other candidate for FOSS explorer.

so... here we are :)

be sure to share any recommendation for windows explorer alternates.

**Reply by GitHubRulesOK** [reply]
I was spoilt as a child by Xtree so nothing better since, however have had to use windows since the day Dos6 died its certainly slow to get used to Windows3 over 30 decades and now the printer explorer has been bugbeared about in WinDos11 too, but that's progress (NOT)

**Reply by tactilis** [reply]
>  "...over 30 decades"

Wow. You _are_ _**old**_.    😉

**Reply by goyalyashpal** [reply]
I will look into Xtree, do you know any FOSS clone/alternate to it? 

searching now on alternativeto:
https://alternativeto.net/software/ztreewin/?license=opensource

Also, can you answer my question at other thread: https://github.com/sumatrapdfreader/sumatrapdf/discussions/3077#discussioncomment-4052840 ??

**Reply by GitHubRulesOK** [reply]
> > "...over 30 decades"
> 
> Wow. You _are_ _**old**_. 😉

yes like DOS well past EOL am writing from my Egyptian Tomb :-) I used paper tape and punch cards when starting with computers they cost mort than a house (portable was a trunk needing a truck to be luggable) and we had to learn the hard ware way.

**Reply by goyalyashpal** [reply]
hey @GitHubRulesOK 

These TUI file explorers or commanders: NNN and Midnight commanders are available on windows via <tt>[msys2]</tt>. I think these may fit your need. Do try them out and share ur experiences :smiley:

After u have installed msys2, here's how to install these packages:
- <tt>pacman -Sy [nnn]</tt>
- <tt>pacman -Sy [mc]</tt>

[msys2]: https://www.msys2.org
[nnn]: https://www.ghacks.net/2019/11/01/nnn-is-an-excellent-command-line-based-file-manager-for-linux-macos-and-bsds/ "via 'user plugins' citation on wikipedia/wiki/nnn"
[mc]: https://en.wikipedia.org/wiki/Midnight_Commander

**Comment by goyalyashpal** [comment]
So, can i change the program used (currently explorer) in `File > Show in folder` or `Ctrl K > Show File In Folder...` to something else _fantasizing thinkin_

**Comment by kjk** [comment, marked as answer]
@yashpalgoyal1304 Could you expand on your use case i.e. what that "something else" is and why would it be better than `Show File In Folder...`

Here are some possible soloutions:

1. We have notion of external viewers https://www.sumatrapdfreader.org/docs/Customize-external-viewers

If I added e.g. %d for "directory in which current file is", you could configure external viewer that is some kind of file manager and open it in the directory

2. Currently commands don't take arguments but I could extend command system to take arguments, add a "exec" command and a way to specify file name, file directory etc. It would be like external viewer but via commands

3. notepad2 has it's own mini-file explorer (metapath) which opens files in notepad2, which is more efficient that traditional File / Open. I've been thinking about doing something similar for sumatra.

In fact, I could implement all 3 above.

But I would like to understand your use case first.

**Reply by goyalyashpal** [reply]
hi! thanks a lot for such a detailed reply. and considering my suggestion.

> you could configure external viewer that is some kind of file manager

1. basically yeah, i want to open the current directory in some other file manager 
specifically: <tt>[doublecmd]</tt> for now, <sup>([commandline] documentation for myself)</sup> 
as i don't want to use the stupid slow and dumb Microsoft Explorer.

[doublecmd]: https://github.com/doublecmd/doublecmd
[commandline]: https://github.com/doublecmd/doublecmd.github.io/blob/master/doc/en/commandline.html

2. thinking of commands, following code block (for external viewer config) worked 😆
but it doesn't populate in the command palette. 


```
	[
		CommandLine = "doublecmd.exe" "%1\.."
		Name = doublecmd
	]
```

3. umh, i don't quite understand the second point on commands with arguments 😅

> 3\. notepad2 has it's own mini-file explorer (metapath)

4. yeah, i'd _guess_ that will be a nice addition too - will _probably_ suite more people... 
but personally, my preference will be the 1st point

----


just for records, 
<p>- notepad2 has a successor: <tt>rizonsesoft/<a href="https://github.com/rizonesoft/Notepad3/tree/master/minipath">notepad3</a></tt> - maybe that'd help you in getting source or latest version</p>
<p>- the minipath in notepad3 can be accessed by "File > Browse" <tt>Ctrl + m</tt>. The "File > Launch > Open with..." <tt>Alt + l</tt> is also <i>probably</i> same</p>

<img src="https://user-images.githubusercontent.com/19423063/218299623-954dfc20-a2b4-4ee6-8f8c-bcf08ef5f7c0.png" width=500px/>

**Comment by kjk** [comment]
I just implemented 1) i.e. in latest [pre-release](https://www.sumatrapdfreader.org/prerelease) (available in about an hour) you can use `%d` to mean `directory in which file is located`.

I've also added built-in `open directory in explorer` external viewer.

I could also add auto-detection of popular file managers (including total commander) as built-ins,  so that you don't have to configure them manually and they would show up in command palette.

**Reply by goyalyashpal** [reply]
https://github.com/sumatrapdfreader/sumatrapdf/commit/76b9e5f35e62d6331675024107d805b0b7b74db6

:+1:

**Comment by kjk** [comment]
Added auto-detection of Total Commander and Directory Opus and their commands so now you can use Command Palette to invoke them.

Assumes they're installed in the same directory as on my machine, which is `"C:\Program Files\totalcmd\TOTALCMD64.EXE"`

**Comment by goyalyashpal** [comment]
with 1.4k ⭐ stars on github, is [doublecmd] popular enough? for introduction, it's native (as against java or electron), lightweight, FOSS, cross platform, keyboard oriented, and orthodox...

[doublecmd]: https://github.com/doublecmd/doublecmd/

i am not really good with judging popularity - i mean, we all do have some kind of bubble right 😅. so, just asking ur opinion!

**Comment by kjk** [comment]
Actually, I mis-read total vs. double in your post.

Added support for Double Commander as well.

Does Double Commander has a `open in existing window` command-line arg and should I add it?

Currently it'll launch a new window every time, which might not be the best option.

**Reply by goyalyashpal** [reply]
> Currently it'll launch a new window every time, which might not be the best option.

yeah, i was thinking same.... tried searching the site for the documentation - stumbled upon the above mentioned source, but couldn't track what's it destination site 😅 

the doublecmd website seems to be split in 3 parts

i am trying to find it, will update with any conclusive result

**Reply by goyalyashpal** [reply]
yeah, found it:

```
-C or --client | If Double Commander is already running, activate it and pass the path(s) in the command line to that instance.
```

the source i stumbled upon https://github.com/sumatrapdfreader/sumatrapdf/discussions/3178#discussioncomment-4944353 was correct - it's destination is this: https://doublecmd.github.io/doc/en/commandline.html

**Reply by kjk** [reply]
Thanks, added `--client` option to double commander, seems to work.

**Comment by GitHubRulesOK** [comment]
@kjk 
many way for external cmd to split old %1filepath or use old %page# so more useful has always been
how to send current word or text $election to local application not just via external web selection 
the icing would be current X&Y or selection xywh

**Comment by notioncolors** [comment]


**Reply by goyalyashpal** [reply]
what? hi spammer :wave:

**Reply by tactilis** [reply]
> I'm just promoting my tools

Don't. You are spamming a SumatraPDF Reader topic that has nothing to do with Notion.

**Reply by shohaghawlader** [reply]
How to remove this post.

**Comment by GitHubRulesOK** [comment]
@goyalyashpal  @tactilis  @shohag-hawlader
thanks 
should now be marked as spam  ?
I think only @kjk has a higher authority to remove parts of thread
